protocol-units/bridge/chains/ethereum/src/lib.rs

```diff
@@ -36,79 +38,85 @@ pub enum Transaction<A, H> {
 	Counterparty(CounterpartyCall<A, H>),
 }
 
-pub struct EthereumChain<A, H, R> {
+pub struct EthereumChain {
 	pub name: String,
 	pub time: u64,
-	pub accounts: HashMap<A, Amount>,
-	pub events: Vec<EthChainEvent<A, H>>,
+	pub accounts: HashMap<EthAddress, Amount>,
+	pub events: Vec<EthChainEvent<EthAddress, EthHash>>,
 
-	pub initiator_contract: SmartContractInitiator<A, H, R>,
-	pub counterparty_contract: SmartContractCounterparty<A, H>,
+	pub initiator_contract: EthClient,
+	pub initiator_monitoring: EthInitiatorMonitoring<EthAddress, EthHash>,
+	pub counterparty_contract: EthClient,
+	pub counterparty_monitoring: EthCounterpartyMonitoring<EthAddress, EthHash>,
 
-	pub transaction_sender: mpsc::UnboundedSender<Transaction<A, H>>,
-	pub transaction_receiver: mpsc::UnboundedReceiver<Transaction<A, H>>,
+	pub transaction_sender: mpsc::UnboundedSender<Transaction<EthAddress, EthHash>>,
+	pub transaction_receiver: mpsc::UnboundedReceiver<Transaction<EthAddress, EthHash>>,
 
-	pub event_listeners: Vec<mpsc::UnboundedSender<EthChainEvent<A, H>>>,
+	pub event_listeners: Vec<mpsc::UnboundedSender<EthChainEvent<EthAddress, EthHash>>>,
 
 	waker: AtomicWaker,
-
-	pub _phantom: std::marker::PhantomData<H>,
+	pub _phantom: std::marker::PhantomData<EthHash>,
 }
 
-impl<A, H, R> EthereumChain<A, H, R>
-where
-	A: BridgeAddressType + From<RecipientAddress<A>>,
-	H: BridgeHashType + GenUniqueHash,
-	R: RngSeededClone,
-	H: From<HashLockPreImage>,
-{
-	pub fn new(mut rng: R, name: impl Into<String>) -> Self {
+impl EthereumChain {
+	pub async fn new(name: impl Into<String>, rpc_url: &str) -> Self {
 		let accounts = HashMap::new();
 		let events = Vec::new();
-		let (event_sender, event_receiver) = mpsc::unbounded();
+		let (_, event_receiver_1) = mpsc::unbounded();
+		let (_, event_receiver_2) = mpsc::unbounded();
```

Comment:
If you drop the sender, the listener won't ever get a message because you can't send them.
I've corrected it in the PR #624. I remove this part and correct in the client side. See previous message